velox/core/PlanNode.cpp

```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
     types.emplace_back(variable->type());
   }
 
+  auto indexCheck = [](vector_size_t index, vector_size_t size) {
+    VELOX_USER_CHECK_LE(
+        index, size, "unnestNames do not include all UnnestNode columns");
+  };
+
   int unnestIndex = 0;
+  auto unnestNamesLength = unnestNames.size();
   for (const auto& variable : unnestVariables_) {
-    if (variable->type()->isArray()) {
-      names.emplace_back(unnestNames[unnestIndex++]);
-      types.emplace_back(variable->type()->asArray().elementType());
-    } else if (variable->type()->isMap()) {
-      const auto& mapType = variable->type()->asMap();
-
+    auto type = variable->type();
+    if (type->isArray()) {
+      auto elementType = type->asArray().elementType();
+      if (elementType->isRow() && !legacyUnnest_) {
+        // Array of rows are flattened to a list of columns. There is one
+        // column for each child type of the row.
+        // Note : In legacy unnest however, array of rows like other arrays
+        // are flattened to a single column of the array type.
+        auto rowType = asRowType(elementType);
```

Comment:
not need for asRowType; Type base class have size() and childAt() APIs, hence, you can use elementType directly.